apps/prairielearn/elements/pl-external-grader-results/pl-external-grader-results.py

```diff
@@ -39,129 +41,140 @@ def ansi_to_html(output):
+    # Early-exit if not the submission panel
+    if data["panel"] != "submission":
+        return ""
+
+    feedback = data["feedback"]
+    results = feedback.get("results", None)
+
+    format_errors = data.get("format_errors", {})
+    grader_format_errors = format_errors.get("_external_grader", [])
+
+    grading_succeeded = bool(feedback.get("succeeded", None))
+
+    html_params = {
+        "submission": True,
+        "graded": bool(feedback),
+        "gradable": results["gradable"]
+        if (results is not None and "gradable" in results)
+        else True,
+        "grading_succeeded": grading_succeeded,
+        "format_errors": grader_format_errors,
+        "has_format_errors": len(grader_format_errors) > 0,
+        "uuid": pl.get_uuid(),
+    }
+
+    if grading_succeeded and results:
+        html_params["succeeded"] = bool(results.get("succeeded", None))
+        html_params["score"] = round_value(results.get("score", 0) * 100)
+        html_params["achieved_max_points"] = results.get("score", 0) >= 1.0
+        html_params["results_color"] = (
+            PLColor("correct_green")
+            if (results.get("score", 0) >= 1.0)
+            else PLColor("incorrect_red")
+        )
+
+        html_params["message"] = ansi_to_html(results.get("message", None))
+        html_params["has_message"] = bool(html_params["message"])
+
+        html_params["output"] = ansi_to_html(results.get("output", None))
+        html_params["has_output"] = bool(html_params["output"])
+
+        html_params["images"] = results.get("images", [])
+        html_params["has_images"] = bool(html_params["images"])
+
+        html_params["has_message_or_output_or_image"] = (
+            html_params["has_message"]
+            or html_params["has_output"]
+            or html_params["has_images"]
+        )
+
+        results_tests = results.get("tests", None)
+        html_params["has_tests"] = bool(results_tests)
+
+        if results_tests:
+            # Let's not assume that people give us a valid array of tests
+            # If any test is missing either points or max_points, we'll
+            # disable detailed scores for all questions
+            tests_missing_points = any(
+                [
+                    test.get("points", None) is None
+                    or test.get("max_points", None) is None
+                    for test in results_tests
+                ]
+            )
+            html_params["tests_missing_points"] = tests_missing_points
+
+            if not tests_missing_points:
+                points_sum = sum(test["points"] for test in results_tests)
+                html_params["points"] = round_value(points_sum)
+
+                max_points_sum = sum(test["max_points"] for test in results_tests)
+                html_params["max_points"] = round_value(max_points_sum)
+
+            # We need to build a new tests array to massage data a bit
+            Test = dict[str, str | bool | int | PLColor | None]
```

Comment:
Put custom type definitions at the top of the file. I think it's ok to use a `dict[str, Any]` here, since we aren't reading these keys again (just passing them to the mustache template). If you want to be very pedantic, you could use a `TypedDict`.

Reply:
I think `TypedDict` is overkill; `dict[str, Any]` makes a lot of sense, I'll go for that.